File: fixtures/canonical-bytes/crossrun/python/runner.py
#!/usr/bin/env python3
# Copyright 2026 Tymofii Pidlisnyi. Apache-2.0 license. See LICENSE.
#
# Python runner for the RFC 8785 canonical-byte cross-run.
#
# Reports where this canonicalizer's bytes and SHA-256 agree with the pinned
# fixture. It is not a verdict on the implementation, and it is not APS
# conformance; it is a byte diff on ten cases.
#
# Canonicalizer under test, in order of preference:
#
#   1. agent_passport.canonical.canonicalize_jcs from the published
#      agent-passport-system distribution on PyPI (see requirements.txt).
#   2. json.dumps(sort_keys=True, separators=(",", ":"), ensure_ascii=False),
#      the standard library encoder, labeled baseline_json_encoder. It does not
#      claim RFC 8785, so its divergences are comparative evidence about the
#      standard library, not a failed conformance run.
#
# WHY (1) IS REFUSED WHEN IT IS AN EDITABLE INSTALL. A `pip install -e` of a
# local checkout resolves to a working tree on this machine, which would make
# the report depend on a sibling repository that a person cloning this suite
# does not have. That is the failure the self-containment rule exists to catch,
# and it is invisible at the import statement, so it is checked explicitly here
# and the runner drops to the labeled baseline instead of reporting a version it
# cannot stand behind.
#
# Every expected value is read from the fixture at run time. Nothing about the
# ten cases is transcribed into this file.
#
# Usage:
#   python3 runner.py [fixture-path]
# Default fixture: ../../canonical-bytes-jcs-v2.json

import hashlib
import json
import os
import platform
import sys
from pathlib import Path

DEFAULT_FIXTURE = Path(__file__).resolve().parent.parent.parent / "canonical-bytes-jcs-v2.json"


def _first_party():
    """Return (fn, name, version) for the published distribution, or None.

    None means either that the distribution is absent or that it resolves to a
    local checkout, which this runner will not report as a published version.
    """
    try:
        import importlib.metadata as md

        from agent_passport.canonical import canonicalize_jcs
    except Exception:
        return None
    try:
        dist = md.distribution("agent-passport-system")
        version = dist.version
        raw = dist.read_text("direct_url.json")
    except Exception:
        return None
    if raw:
        try:
            info = json.loads(raw)
        except ValueError:
            return None
        # A direct_url pointing at a directory, editable or not, is a local
        # checkout rather than an artifact resolved from an index.
        if info.get("dir_info") is not None or str(info.get("url", "")).startswith("file://"):
            return None
    return canonicalize_jcs, "agent_passport.canonical.canonicalize_jcs", version


def _baseline(obj):
    return json.dumps(obj, sort_keys=True, separators=(",", ":"), ensure_ascii=False)


def first_divergent_byte_offset(a: bytes, b: bytes):
    """Zero-based offset of the first differing byte.

    When one sequence is an exact prefix of the other there is no differing
    byte, so the answer is the length of the shorter one: the offset a reader
    would look at to see where the two stopped agreeing.
    """
    shared = min(len(a), len(b))
    for i in range(shared):
        if a[i] != b[i]:
            return i
    return None if len(a) == len(b) else shared


def main() -> int:
    fixture_path = Path(sys.argv[1]).resolve() if len(sys.argv) > 1 else DEFAULT_FIXTURE
    fixture_bytes = fixture_path.read_bytes()
    fixture = json.loads(fixture_bytes)

    chosen = _first_party()
    if chosen is not None:
        canonicalize, implementation, version = chosen
        kind = "first_party"
    else:
        canonicalize = _baseline
        implementation = 'json.dumps(sort_keys=True, separators=(",",":"), ensure_ascii=False)'
        version = platform.python_version()
        kind = "baseline_json_encoder"
        # Diagnostic on stderr, so stdout stays a clean JSON document. Without
        # this a reader sees "baseline_json_encoder" and cannot tell whether the
        # package is missing, unusable, or was refused for being a local
        # checkout, which are three different things to do something about.
        print(
            "note: falling back to the standard library encoder. "
            "Install the published distribution for a first-party result: "
            "python3 -m pip install -r " + str(Path(__file__).resolve().parent / "requirements.txt"),
            file=sys.stderr,
        )

    cases = []
    for vector in fixture["vectors"]:
        actual = canonicalize(vector["input"]).encode("utf-8")
        expected = bytes.fromhex(vector["canonical_bytes_hex"])
        actual_sha = hashlib.sha256(actual).hexdigest()
        byte_match = actual == expected
        cases.append({
            "name": vector["name"],
            "byte_match": byte_match,
            "sha256_match": actual_sha == vector["canonical_sha256"],
            "actual_bytes_hex": actual.hex(),
            "actual_sha256": actual_sha,
            "first_divergent_byte_offset": None if byte_match else first_divergent_byte_offset(actual, expected),
        })

    json.dump({
        "runner": "python",
        "implementation": implementation,
        "implementation_kind": kind,
        "implementation_version": version,
        "runtime_version": f"python {platform.python_version()}",
        "fixture": str(fixture_path),
        "fixture_sha256": hashlib.sha256(fixture_bytes).hexdigest(),
        "cases": cases,
        "summary": {
            "total": len(cases),
            "byte_match": sum(1 for c in cases if c["byte_match"]),
            "sha256_match": sum(1 for c in cases if c["sha256_match"]),
        },
    }, sys.stdout, indent=2)
    sys.stdout.write("\n")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: fixtures/canonical-bytes/crossrun/ts/runner.ts
// Copyright 2026 Tymofii Pidlisnyi. Apache-2.0 license. See LICENSE.
//
// TypeScript runner for the RFC 8785 canonical-byte cross-run.
//
// Reports where this canonicalizer's bytes and SHA-256 agree with the pinned
// fixture. It is not a verdict on the implementation, and it is not APS
// conformance; it is a byte diff on ten cases.
//
// Canonicalizer under test: canonicalizeJCS from the published
// agent-passport-system npm package, resolved as a bare specifier so a clean
// checkout plus `npm ci` is self-contained. No sibling checkout, no $HOME, no
// path dependency.
//
// Every expected value is read from the fixture at run time. Nothing about the
// ten cases is transcribed into this file: a runner that embedded a byte string
// would still agree with itself after the fixture changed.
//
// Usage:
//   npx tsx runner.ts [fixture-path]
// Default fixture: ../../canonical-bytes-jcs-v2.json

import { createHash } from 'node:crypto'
import { readFileSync } from 'node:fs'
import { dirname, join, resolve } from 'node:path'
import { fileURLToPath } from 'node:url'

import { canonicalizeJCS } from 'agent-passport-system'

const __dirname = dirname(fileURLToPath(import.meta.url))
const DEFAULT_FIXTURE = join(__dirname, '..', '..', 'canonical-bytes-jcs-v2.json')

const fixturePath = resolve(process.argv[2] ?? DEFAULT_FIXTURE)
const fixtureBytes = readFileSync(fixturePath)
const fixture = JSON.parse(fixtureBytes.toString('utf8')) as {
  vectors: { name: string; input: unknown; canonical_bytes_hex: string; canonical_sha256: string }[]
}

const sha256Hex = (buf: Buffer): string => createHash('sha256').update(buf).digest('hex')

/** Zero-based offset of the first differing byte. When one sequence is an exact
 *  prefix of the other there is no differing byte at all, so the answer is the
 *  length of the shorter one: that is the offset a reader would look at to see
 *  where the two stopped agreeing. Null when the bytes are equal. */
function firstDivergentByteOffset(a: Buffer, b: Buffer): number | null {
  const shared = Math.min(a.length, b.length)
  for (let i = 0; i < shared; i++) if (a[i] !== b[i]) return i
  return a.length === b.length ? null : shared
}

// The version is read from the manifest of the package that ACTUALLY loaded,
// rather than typed here or read from this repo's dependency pin, so it cannot
// drift from what was imported. Two details make this less direct than it looks:
// the package's exports map declares no ./package.json subpath, so the manifest
// cannot be imported, and it declares only an `import` condition, so
// require.resolve cannot see it either. import.meta.resolve is the ESM-native
// resolver that does, and the manifest is then found by walking up from the
// resolved entry point. The walk stays inside the resolved package; it never
// consults $HOME or a sibling checkout.
function loadedPackageVersion(specifier: string): string {
  let dir = dirname(fileURLToPath(import.meta.resolve(specifier)))
  for (;;) {
    const candidate = join(dir, 'package.json')
    try {
      const manifest = JSON.parse(readFileSync(candidate, 'utf8')) as { name?: string; version?: string }
      if (manifest.name === specifier && manifest.version) return manifest.version
    } catch { /* keep walking */ }
    const parent = dirname(dir)
    if (parent === dir) throw new Error(`could not find the manifest of ${specifier}`)
    dir = parent
  }
}
const sdkVersion = loadedPackageVersion('agent-passport-system')

const cases = fixture.vectors.map(v => {
  const actual = Buffer.from(canonicalizeJCS(v.input), 'utf8')
  const expected = Buffer.from(v.canonical_bytes_hex, 'hex')
  const actualSha = sha256Hex(actual)
  const byteMatch = actual.equals(expected)
  return {
    name: v.name,
    byte_match: byteMatch,
    sha256_match: actualSha === v.canonical_sha256,
    actual_bytes_hex: actual.toString('hex'),
    actual_sha256: actualSha,
    first_divergent_byte_offset: byteMatch ? null : firstDivergentByteOffset(actual, expected),
  }
})

process.stdout.write(JSON.stringify({
  runner: 'ts',
  implementation: 'agent-passport-system canonicalizeJCS',
  implementation_kind: 'first_party',
  implementation_version: sdkVersion,
  runtime_version: `node ${process.versions.node}`,
  fixture: fixturePath,
  fixture_sha256: sha256Hex(fixtureBytes),
  cases,
  summary: {
    total: cases.length,
    byte_match: cases.filter(c => c.byte_match).length,
    sha256_match: cases.filter(c => c.sha256_match).length,
  },
}, null, 2) + '\n')
